packages/arb-checkpointer/checkpointing/checkpointedmachine.go

```diff
@@ -0,0 +1,255 @@
+	"github.com/offchainlabs/arbitrum/packages/arb-checkpointer/ckptcontext"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/common"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/machine"
+	"github.com/offchainlabs/arbitrum/packages/arb-util/value"
+	"github.com/offchainlabs/arbitrum/packages/arb-validator-core/arbbridge"
+)
+
+var errNoCheckpoint = errors.New("cannot restore because no checkpoint exists")
+var errNoMatchingCheckpoint = errors.New("cannot restore because no matching checkpoint exists")
+
+type CheckpointedMachine struct {
+	*sync.Mutex
+	cm                    *cmachine.CheckpointedMachine
+	nextCheckpointToWrite *writableCheckpoint
+	maxReorgHeight        *big.Int
+}
+
+func NewCheckpointedMachine(
+	rollupAddr common.Address,
+	executablePath string,
+	databasePath string,
+	maxReorgHeight *big.Int,
+	forceFreshStart bool,
+) (*CheckpointedMachine, error) {
+	ret, err := newCheckpointedMachine(
+		rollupAddr,
+		executablePath,
+		databasePath,
+		new(big.Int).Set(maxReorgHeight),
+		forceFreshStart,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	go ret.writeDaemon()
+	// TODO
+	//go cleanupDaemon(ret.bs, ret.db, maxReorgHeight)
+	return ret, nil
+}
+
+// newCheckpointedMachine creates the checkpointedmachine, but doesn't
+// launch it's reading and writing threads. This is useful for deterministic
+// testing
+func newCheckpointedMachine(
+	rollupAddr common.Address,
+	executablePath string,
+	databasePath string,
+	maxReorgHeight *big.Int,
+	forceFreshStart bool,
+) (*CheckpointedMachine, error) {
+	if databasePath == "" {
+		databasePath = MakeCheckpointDatabasePath(rollupAddr)
+	}
+	if forceFreshStart {
+		// for testing only --  delete old database to get fresh start
+		if err := os.RemoveAll(databasePath); err != nil {
+			return nil, err
+		}
+	}
+	cCheckpointedMachine, err := cmachine.NewCheckpointedMachine(executablePath, databasePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &CheckpointedMachine{
+		new(sync.Mutex),
+		cCheckpointedMachine,
+		nil,
+		maxReorgHeight,
+	}, nil
+}
+
+func (cp *CheckpointedMachine) Initialize(arbitrumCodeFilePath string) error {
+	return cp.cm.Initialize(arbitrumCodeFilePath)
+}
+
+func (cp *CheckpointedMachine) Initialized() bool {
+	return cp.cm.Initialized()
+}
+
+func (cp *CheckpointedMachine) MaxReorgHeight() *big.Int {
+	return new(big.Int).Set(cp.maxReorgHeight)
+}
+
+func (cp *CheckpointedMachine) GetAggregatorStore() *cmachine.AggregatorStore {
+	return cp.cm.GetAggregatorStore()
+}
+
+// HasCheckpointedState checks whether the block store is empty, which is the table
+// which contains the checkpoints recorded by the CheckpointedMachine
+func (cp *CheckpointedMachine) HasCheckpointedState() bool {
+	// TODO
+	return true
+}
+
+func (cp *CheckpointedMachine) GetInitialMachine(vc machine.ValueCache) (machine.Machine, error) {
+	return cp.cm.GetInitialMachine(vc)
+}
+
+func (cp *CheckpointedMachine) AsyncSaveCheckpoint(
```

Comment:
The old checkpointer was designed to replace pending checkpoints with new ones if another save is requested before the previous beings. We should confirm that the behavior still makes sense